File: scratch.m


%% Repeating Patterns
% Define a vector that repeats n times, such that:
x = repmat([0, 1, 2, 3]', 9, 1);

% then try to model it using the forward difference approach
x0 = [0, 1, 2, 3, 0];

function dx = forward_diff(x, ii, order)
    % evaluates the order'th derivative at index ii of the vector x
    assert(ii > order, 'length of x must be greater than derivative order')

    if order == 0
        dx = x(ii); return;
    elseif order == 1
        dx = -x(ii) + x(ii-1); return;
    elseif order == 2
        dx = x(ii) - 2*x(ii-1) x(ii-2); return;
    elseif order == 3
        dx = -x(ii) + 3*x(ii-1) - 3*x(ii-2) + x(ii-3); return;
    elseif order == 4
        dx = x(ii) - 4*x(ii-1) + 6*x(ii-2) - 4*x(ii-3) + x(ii-4); return;
    end
end

function y = iterate(x, t)
    % given the vector x, predict the rule and iterate forward t steps
    % using the Hille series
    dx1 = forward_diff(x, length(x), 1);
    dx2 = forward_diff(x, length(x), 2);
    dx3 = forward_diff(x, length(x), 3);
    dx4 = forward_diff(x, length(x), 4);

    y = zeros(length(x) + t, 1);
    y(1:length(x)) = x;

end

File: scratch.jl
## Attempt to track a repeating trajectory using the Hille series

# preamble
using LinearAlgebra

# initial vector
x₀ = [0, 1, 2, 3, 0, 1]

# repeat the vector n times
x = repeat(x₀, 3)

# define the forward difference
function forward_diff(x)
    # compute the forward difference of x at ii
    # up to fourth order differences
    @assert length(x) > 5 "length(x) must be > 5"

    coeffs = [-1 1 0 0 0;
              1 -2 1 0 0;
              -1 3 -3 1 0;
              1 -4 6 -4 1]

    dx = [coeffs[i, :] ⋅ x[end-4:end] for i ∈ 1:4]
    return Tuple(dx)
end

# define the Hille series trajectory iteration
function evolve(x, t)
    # iterate x using the forward difference t steps
    y = zeros(length(x) + t)
    dx = forward_diff(x)

    function hille(x, dx, t)
        value = x + t * dx[1] + (1/2)t^2*dx[2] + (1/6)t^3*dx[3] + (1/24)t^4*dx[4]
    end

    for i ∈ 1:t
        y[length(x)+i] = hille(x[end], dx, i)
    end

    return y
end
